pulp_container/app/tasks/sync_stages.py

```diff
@@ -123,13 +127,40 @@ async def run(self):
         ) as pb_parsed_tags:
 
             for download_tag in asyncio.as_completed(to_download):
-                dl_res = await download_tag
-                with open(dl_res.path, "rb") as content_file:
-                    raw_data = content_file.read()
-                dl_res.artifact_attributes["file"] = dl_res.path
-                saved_artifact = await sync_to_async(_save_artifact_blocking)(
-                    dl_res.artifact_attributes
-                )
+                async with ProgressReport(
+                    message="Processing Manifests", code="sync.processing.manifest"
+                ) as pr_manifest:
+                    dl_res = await download_tag
+
+                    digest = dl_res.headers.get("docker-content-digest")
+                    manifest = await sync_to_async(
+                        Manifest.objects.prefetch_related("contentartifact_set")
+                        .filter(digest=digest)
+                        .first
+                    )()
+                    if manifest:
+
+                        def _get_content_data_blocking():
+                            saved_artifact = manifest.contentartifact_set.first().artifact
+                            content_file = saved_artifact.file
+                            raw_data = content_file.read()
+                            return saved_artifact, raw_data
+
+                        saved_artifact, raw_data = await sync_to_async(_get_content_data_blocking)()
+
+                    else:
+                        await pr_manifest.aincrement()
+                        downloader = self.remote.get_downloader(url=dl_res.url)
+                        dl_res = await downloader.run(extra_data={"headers": V2_ACCEPT_HEADERS})
```

Comment:
Are we losing some concurrency here by `await`ing each `run()` co-routine instead of collecting them and then scheduling them all at once using `asyncio.as_completed()` ?

Reply:
Yes we are. To make it more concurrently we probably gonna need to break it into smaller and coherent pieces.

Reply:
we can probably take similar approach as in  here with listed manifests https://github.com/pulp/pulp_container/blob/main/pulp_container/app/tasks/sync_stages.py#L146